src/_pytest/python.py

```diff
@@ -1190,14 +1190,25 @@ def idmaker(argnames, parametersets, idfn=None, ids=None, config=None, item=None
         _idvalset(valindex, parameterset, argnames, idfn, ids, config=config, item=item)
         for valindex, parameterset in enumerate(parametersets)
     ]
-    if len(set(ids)) != len(ids):
-        # The ids are not unique
-        duplicates = [testid for testid in ids if ids.count(testid) > 1]
-        counters = Counter()
+
+    # All ids must be unique!
+    unique_ids = set(ids)
+    if len(unique_ids) != len(ids):
+
+        # Record the number of occurances of each test id
+        testid_counts = Counter(ids)
+
+        # Map the test id to its next suffix.
+        testid_suffixes = Counter(unique_ids)
+        for testid in testid_suffixes.keys():
+            testid_suffixes[testid] -= 1  # start each suffix at 0
+
+        # Suffix non-unique ids to make them unique:
         for index, testid in enumerate(ids):
-            if testid in duplicates:
-                ids[index] = testid + str(counters[testid])
-                counters[testid] += 1
+            if testid_counts[testid] > 1:
+                ids[index] = f"{testid}{testid_suffixes[testid]}"
```

Comment:
Python 3.5 is supported so unfortunately we can't use f-strings yet.
```suggestion
                ids[index] = "{}{}".format(testid, testid_suffixes[testid])
```